utils/src/main/java/software/amazon/awssdk/utils/cache/CachedSupplier.java

```diff
@@ -156,54 +188,117 @@ private void refreshCache() {
+                    try {
+                        cachedValue = handleFetchedSuccess(prefetchStrategy.fetch(valueSupplier));
+                    } catch (RuntimeException t) {
+                        cachedValue = handleFetchFailure(t);
+                    }
                 }
             } finally {
                 if (lockAcquired) {
                     refreshLock.unlock();
                 }
             }
         } catch (InterruptedException e) {
-            handleInterruptedException("Interrupted waiting to refresh the value.", e);
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException("Interrupted waiting to refresh a cached value.", e);
         }
     }
 
-    private void handleInterruptedException(String message, InterruptedException cause) {
-        Thread.currentThread().interrupt();
-        throw new IllegalStateException(message, cause);
+    /**
+     * Perform necessary transformations of the successfully-fetched value based on the stale value behavior of this supplier.
+     */
+    private RefreshResult<T> handleFetchedSuccess(RefreshResult<T> fetch) {
+        consecutiveStaleRetrievalFailures.set(0);
+
+        Instant now = clock.instant();
+
+        if (now.isBefore(fetch.staleTime())) {
+            return fetch;
+        }
+
+        switch (staleValueBehavior) {
+            case STRICT:
+                Instant newStale = now.plusSeconds(1);
+                log.warn(() -> "Retrieved value expiration is in the past (" + fetch.staleTime() + "). Using expiration "
+                               + "of " + newStale);
+                return fetch.toBuilder().staleTime(newStale).build(); // Refresh again in 1 second
+            case ALLOW:
+                Instant newStaleTime = jitterTime(now, Duration.ofMinutes(1), Duration.ofMinutes(10));
+                log.warn(() -> "Cached value expiration has been extended to " + newStaleTime + " because the downstream "
+                               + "service returned a time in the past: " + fetch.staleTime());
+
+                return fetch.toBuilder()
+                            .staleTime(newStaleTime)
+                            .build();
+            default:
+                throw new IllegalStateException("Unknown stale-value-behavior: " + staleValueBehavior);
+        }
+    }
+
+    /**
+     * Perform necessary transformations of the currently-cached value based on the stale value behavior of this supplier.
+     */
+    private RefreshResult<T> handleFetchFailure(RuntimeException e) {
+        RefreshResult<T> currentCachedValue = cachedValue;
+        if (currentCachedValue == null) {
+            throw e;
+        }
+
+        Instant now = clock.instant();
+        if (!now.isBefore(currentCachedValue.staleTime())) {
+            int numFailures = consecutiveStaleRetrievalFailures.incrementAndGet();
+
+            switch (staleValueBehavior) {
+                case STRICT:
+                    throw e;
```

Comment:
We failed to fetch from the underlying supplier and the current value is stale. Throw that exception up to the caller. This is a slight behavior change from what exists today, where we'd throw the exception even if the current value wasn't stale (i.e. it was a prefetch). This behavior felt more friendly.